src/router.rs

```diff
@@ -32,7 +36,7 @@ impl<'a> Router<'a> {
             return Err("UUID already exists".to_string());
         }
 
-        let ast = parse(atc).map_err(|e| e.to_string())?;
+        let ast = parse(atc, &mut self.regex_cache).map_err(|e| e.to_string())?;
```

Comment:
Nice work, however, it seems that we never check to see if `Rc` gets down to `1` and regex should be released. We probably need to check the `Rc` count in `remove`, and if it gets to `1` (which means only `HashMap` is holding it), remove the entry.

Reply:
Huh! I exploring if there's a way to register a hook to `Rc` that will get triggered on `drop` but it does not look like there's anything. But you're right - we don't need something as sophisticated and complicated as that. The action of removing a matcher is a great place (even better as it's explicit) to do such removal. Will do that 👍

Reply:
Fixed ✅

Reply:
Yes, we try to make things as predictable as possible in Rust, so checking it only at the place it should be checked is preferred.